### PR TITLE
[REVIEW] Replace index type-dispatch call with indexalator in cudf::strings::substring

### PR DESCRIPTION
**Pull request by davidwendt**
One of the `cudf::strings::substring` APIs accepts a list of character positions (i.e. indices) for slicing individual strings. The type-dispatcher is used to accept any integer type as the positions column parameters. Since these are indices, the indexalator can be used in place of the type-dispatcher. This will simplify the code and make it build faster. The compile time for `substrings.cu.o` reduced from 70s to 46s.

### CODE REVIEW

File: cpp/src/strings/substring.cu [hunk 1]
```diff
@@ -168,182 +155,140 @@ std::unique_ptr<column> slice_strings(strings_column_view const& strings,
...
-    return make_strings_column(strings_count,
-                               std::move(offsets_column),
-                               std::move(chars_column),
-                               null_count,
-                               std::move(null_mask),
-                               stream,
-                               mr);
-  }
-};
+std::unique_ptr<column> compute_substrings_from_fn(column_device_view const& d_column,
```
Comment (rgsl888prabhu) [inline cpp/src/strings/substring.cu:187]: It would be nice to have @brief here

**Review by rgsl888prabhu — APPROVED** [review]
LGTM

**Review by nvdbaranec — CHANGES_REQUESTED** [review]
One tiny wording thing.

**Comment by davidwendt** [comment]
rerun tests

**Review by nvdbaranec — APPROVED** [review]
lgtm

**Pull request merged**